common/src/main/java/io/netty/util/NettyRuntime.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.util;
+
+import io.netty.util.internal.SystemPropertyUtil;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A utility class for wrapping calls to {@link Runtime}.
+ */
+public final class NettyRuntime {
+
+    private static final AtomicInteger availableProcessors = new AtomicInteger();
+
+    /**
+     * No public constructor to prevent instances from being created.
+     */
+    private NettyRuntime() {
+    }
+
+    /**
+     * Set the number of available processors.
+     *
+     * @param availableProcessors the number of available processors
+     * @throws IllegalArgumentException if the specified number of available processors is non-positive
+     * @throws IllegalStateException    if the number of available processors is already configured
+     */
+    public static void setAvailableProcessors(final int availableProcessors) {
+        if (availableProcessors <= 0) {
+            throw new IllegalArgumentException(
+                    "availableProcessors must be positive but was [" + availableProcessors + "]");
+        }
+        if (!NettyRuntime.availableProcessors.compareAndSet(0, availableProcessors)) {
+            throw new IllegalStateException(
+                    "availableProcessors is already set to [" + NettyRuntime.availableProcessors.get() + "], " +
+                            "rejecting [" + availableProcessors + "]");
+        }
+    }
+
+    /**
+     * Get the configured number of available processors. The default is {@link Runtime#availableProcessors()}. This
+     * can be overridden by setting the system property "io.netty.availableProcessors" or by invoking
+     * {@link #setAvailableProcessors(int)} before any calls to {@link #availableProcessors()}.
+     *
+     * @return the configured number of available processors
+     */
+    @SuppressForbidden(reason = "to obtain default number of available processors")
+    public static int availableProcessors() {
+        if (NettyRuntime.availableProcessors.get() == 0) {
+            final int availableProcessors =
+                    SystemPropertyUtil.getInt(
+                            "io.netty.availableProcessors",
+                            Runtime.getRuntime().availableProcessors());
+            setAvailableProcessors(availableProcessors);
```

Comment:
> yes different static blocks or it could even be just different methods that call it that run concurrently. Why would you not expect something like this happen ? 

No, I don't expect that. That's why I am acknowledging that potential for a race is there but I am saying that I think it's fine in practice.

For example, look at `NioEventLoopGroup` which extends `MultithreadEventLoopGroup`. There is a class initializer on `MultithreadEventLoopGroup` with this block of code:

```
  static {
        DEFAULT_EVENT_LOOP_THREADS = Math.max(1, SystemPropertyUtil.getInt(
                "io.netty.eventLoopThreads", NettyRuntime.availableProcessors() * 2));

        if (logger.isDebugEnabled()) {
            logger.debug("-Dio.netty.eventLoopThreads: {}", DEFAULT_EVENT_LOOP_THREADS);
        }
    }
```

This is guaranteed to only execute once. It can not be the case that constructing two event loop groups lead to a get race here, the class initializer guards against that.

Since this is used statically everywhere, the only way to get the race that you're concerned about it by being nefarious. 

> Shouldnt it be safe to call a method concurrently that does lazy init ?

The downside to your way is that now you allow a get racing with a set, all silently being fine for a thousand runs until one day an evil scheduler executes the get before the set.

This is why I favor the approach that I took here. I do not think a race arises in practice. My way disallows all races: get with get, and get with set so there will always be a loser that throws an `IllegalStateException` so an application programmer fixes their buggy code and avoids the race to begin with. With your way, you introduce the possibility of a get racing with a set, but only sometimes is there a loser that throws an `IllegalStateException` (if the set wins, everything is fine, if the get wins, the set loses); that's why this I'm opposed to this option.